Fix Re-enabling manual Style creation

mhelsley opened the pull request:
Fixes #46 by moving OSControl out of Style and into AnsiGenericString

fdncred commented on src/display.rs:
```diff
@@ -310,13 +351,39 @@ where
+            Some(OSControl::Title) => {
+                write!(w, "\x01\x1B]2;")?;
+                w.write_str(self.string.as_ref())?;
+                write!(w, "\x1B\x5C\x02")
```
I'm not sure what these ending chars are (\x5C\x02). I thought Title, Icon, Title-with-Icon, and Hyperlink were all supposed to end with BEL (0x07) or ST (0x9C)? Am I misunderstanding?

Also, Title-with-Icon is missing I think
https://tldp.org/HOWTO/Xterm-Title-3.html

Hyperlink info https://gist.github.com/egmontkob/eb114294efbcd5adb1944c9f3cb5feda

mhelsley replied:
My primary reference has been the [control sequences from xterm](https://invisible-island.net/xterm/ctlseqs/ctlseqs.html#h3-Operating-System-Commands) which is also consistent with the references you linked.

> @fdncred
> I'm not sure what these ending chars are (\x5C\x02)

The first byte is "\". The second is second part of a pair that wraps zero-width sequences. So when the terminal sees \x01 ... \x02 it knows the ... portion shouldn't be made visible. Incidentally these are called:
0x01 SOH Start of Heading
0x02 STX Start of Text

I've been testing these sequences in prompt strings and without those zero-width markings it messes up libraries like readline. If you look at the newly emitted sequences in this PR you should see that most start with 0x01 because, with the exception of a portion of the hyperlink, most OSC sequences in this PR are not meant to be displayed in the terminal.

> @fdncred
> I thought Title, Icon, Title-with-Icon, and Hyperlink were all supposed to end with BEL (0x07) or ST (0x9C)? Am I misunderstanding?

You are correct. However 0x9C (notice it's > 0x7F) is part of the "C1" section and from what I've read everything in the C1 section tends to conflict with non-ASCII character encodings (UTF-8 and UTF-16) since they use values above 0x7F. So I've chosen to stick to the "C0" section to maximize portability and/or avoid needing to detect the capabilities and mode of the current terminal.

As for the question of using BEL vs ST, the xterm docs on control sequences state that it supports both for OSC sequences. The PR is using ST which is also used in the hyperlink example found in the gist you linked.

> @fdncred 
> Also, Title-with-Icon is missing I think

Yes, and I omitted it for a couple reasons.

Choosing the specific ones that only do one or the other -- set the icon or title but not both -- is probably more portable. Also it's pretty unclear what it means to set the title and icon name to the same thing at the same time so I figure that ambiguity might be good to avoid engaging with when defining an API.

If users of the API want to set both that can call them separately.

Also, if it proves to be needed in the future the API can still be extended to support that.

[Regarding portability, this TLDP page says:](https://tldp.org/HOWTO/Xterm-Title-6.html)
> "The Irix terminals also support the xterm escapes to individually set window title and icon title, but not the escape to set both."

--
The xterm reference I gave is a bit long so I thought it might help to quote useful parts here.

In the OSC section it says:
> XTerm accepts either BEL  or ST  for terminating OSC

It also helpfully describes the trouble with using C1 codes in modern systems.
> ECMA-48 (aka "ISO 6429") documents C1 (8-bit) and C0 (7-bit) codes.
> Those are respectively codes 128 to 159 and 0 to 31.
...
> Controls (including the escape code 27) are processed once:
> 
> o   This means that a C1 control can be mistaken for badly-formed UTF-8
>      when the terminal runs in UTF-8 mode because C1 controls are valid
>      continuation bytes of a UTF-8 encoded (multibyte) value.

fdncred replied:
Thanks for the thorough explanation, it helps.

I'm all for better portability and generally agree with your choices. However, I'm concerned about 0x01 and 0x02 bits. What precedent to you have for using this? I didn't see them mentioned in the xterm doc you referenced. I've gone through that thing many times myself but may have just missed it.

I have noticed that some terminals respond better to BEL than ST or vice versa. I wonder if SOH/STX would fix those situations. Not asking you to do any testing or anything but your comments made me think about it.

mhelsley replied:
> Thanks for the thorough explanation, it helps.

Great! I tend to look at this as part of documenting the changes so I err on the side of being thorough. It's especially important because some of this is rather obscure and it's taken me a while to dig it up, check it in references for multiple terminal emulators.

> I'm all for better portability and generally agree with your choices. However, I'm concerned about 0x01 and 0x02 bits. What precedent to you have for using this? I didn't see them mentioned in the xterm doc you referenced. I've gone through that thing many times myself but may have just missed it.

Ah, good point. That comes from documentation surrounding prompt generation. It's used indirectly in starship in the form of `\[` and `\]` (Note that the backslashes appear in the strings -- they're not escaping the square brackets) when building prompts for bash. Bash then expands strings, when found in $PS1, into `0x01` and `0x02` respectively ( \#define for  RL_PROMPT_START_IGNORE and RL_PROMPT_END_IGNORE). According to the starship code for zsh and tcsh those shells similarly translate the strings `%{` and `}%` when expanding prompts. That being said, I don't see any way for the terminal emulator to know these sequences are part of shell prompts so I think it treats anything wrapped in 0x01 ... 0x02 as zero-width.

In some documentation they're referred to as `^A` and `^B`

https://en.wikipedia.org/wiki/C0_and_C1_control_codes#SOH

In the bash source code, lib/readline/ there's this (prefixed with line numbers):
```
./readline.h
885:#define RL_PROMPT_START_IGNORE	'\001'

./display.c
341: \001 (^A) start non-visible characters
342:           \002 (^B) end non-visible characters
343:     all characters except \001 and \002 (following a \001) are copied to
344:     the returned string; all characters except those between \001 and
345:     \002 are assumed to be `visible'.
```

In the context of .inputrc (which at least readline uses see: https://medium.com/everything-full-stack/terminal-escape-codes-are-awesome-heres-why-c8eb938b1a1c ):
```
# \1 — Start zero-width character (\001)
# \2 — End zero-width character (\002)
# \e — By itself simply means “an escape character”
# \e[ — Signal CSI (Control Sequence Introducer)
# \e] — Signal OSC (Operating System Command)
```

Also another helpful resource for wezterm (which I've seen references to but never used): https://wezfurlong.org/wezterm/escape-sequences.html#c0-control-codes

suggests these are "ignored".

It's a little unsettling that these aren't *documented* in the xterm references but I have *tested* them under xterm and, as best I can tell, they work exactly as expected. I even looked to see if they're documented in the vt100 references I could find and no such luck.

> I have noticed that some terminals respond better to BEL than ST or vice versa. I wonder if SOH/STX would fix those situations. Not asking you to do any testing or anything but your comments made me think about it.

Hmm, interesting. I've just been manually testing in gnome-terminal, xterm, eterm, rxvt, and alacritty. I only have Linux systems so I can't test iTerm2 or windows terminal software. So far they all seem to support ST but it's also quite likely I've missed some combinations (os + term) and I'm curious to get a sense of what those are if you can recall specific ones I can try.

One interesting related part of  the wezterm reference says:
```
OSC sequences are canonically delimited by the ST (String Terminator) sequence, but WezTerm also accepts delimiting them with the BEL control.
```

fdncred replied:
TIL - All this stuff. Thanks for the education! :)

You last few points hit on where I was going, which is being cross-platform friendly. We forked this crate to update it a little but to also allow it to work in nushell well. Nushell runs on MacOS, Linux, and Windows (amongst others). So, we try our best to support these major 3 platforms everywhere.

Here's a list of terminals I personally try to test with/use off and on, although I'm suer that forgetting a few.

WezTerm - MacOS (still kind of broken in Windows)
Alacritty - MacOS/Linux
Windows Terminal - Windows
gnome terminal - Linux
Terminal.app - MacOS
Tabby - MacOS/Windows
Kitty - MacOS/Linux
iTerm2 - MacOS
Visual Studio Code - MacOS/Linux/Windows

So, we really need to test a few on the platforms that you haven't tested. Any tips on how to do that?

mhelsley replied:
> So, we really need to test a few on the platforms that you haven't tested. Any tips on how to do that?

Well, the way I've been testing is I use a branch of Starship as my command line prompt generator. It tests out the hyperlink bits every time I enter a new command on any of my open terminals.

As for `::title()` and `::cwd()`: I've been testing those sequences more manually by echo-ing the output strings but I'm hoping to add bits to Starship for those as well.

`::icon()` is trickier to test since I'm not aware of anything that pays attention to the "icon title" any more. I think this was used in Ye Olden Days when X11 windows could be minimized into an icon above some text that would be used to restore it. In modern Linux desktops (e.g. Wayland-based rather than X.org) I'm not sure there are corresponding elements to set much less if they ever get displayed somewhere.

One difficulty is a lot of command line prompts set the window title so to test that manually or via cargo you've got to disable that (part of) your prompt generator. So while I've been tempted to write an example testcase much like the table of colors I haven't (yet) thought of a way to temporarily disable the prompt resetting it before you can see the result.

Something like PhantomJS but for terminals sure would be a Nice-to-Have.

fdncred replied:
Do we have tests to cover this that run in the CI? If not, what can we do to test these features in CI? We probably also need to add macos-latest in the CI as well.

mhelsley replied:
OK, I hear you on the CI end of things.

For now I'm trying to make it easier to manually test.

First: I've written some examples much like the `cargo run --example 256_colors` in the current CI runs. The `::title()`, `::icon()`, and `::cwd()` examples sleep for 5 seconds so if you run them you have a moment to look at, reveal , or use those respectively in the terminal emulators that support them.

It's still a manual process, though, and the user needs to know what their shell does with that information. For example, on my machine (Linux, GNOME Shell, gnome-terminal), I can only see the icon title if I enter "overview" mode where I can see the running application icons and icon titles for the apps visible on that screen. However my terminals don't seem to do anything with `::cwd()` yet I've heard that Mac OS Terminal.app uses it when opening a new terminal window from an existing terminal.

So it's still very much a manual test and that's why I put these in the examples.

I'm also looking into optionally running combinations of terminal, shell, and OSC output when users run `cargo test osc`.
For example, on Linux, it could cross gnome-terminal, xterm, rxvt, Eterm, and alacritty with bash, zsh, and tcsh, emit the control sequence for each OSC function, and the tester would see the "output". Since it's starting new terminal emulator windows for each they should appear showing the correct titles (window and icon), working directory, and hyperlinks.

But I still don't see how we could get that into CI hence those tests wouldn't be run with a simple `cargo test` nor `cargo test --all-features`

Before I think about CI integration some more I'm going to get these tests finished and pushed as additional commits so you folks can see them.

fdncred replied:
> The ::title(), ::icon(), and ::cwd() examples sleep for 5 seconds so if you run them you have a moment to look at, reveal , or use those respectively in the terminal emulators that support them.

I think that's fine for one-off testing but not for CI. We just need to ensure that the right escape sequence is being written out. As you know, terminals decide for themselves how to interpret any ansi escape we create.

>  However my terminals don't seem to do anything with ::cwd() ...

When we were working on this in nushell, what I remember about it was that I had to change a directory, for example `cd /tmp`. Then, on terminals that support tabs, I start a new tab, and it should start in `/tmp`. Like you're saying, not all terminals support that escape sequence.

So, reiterating, we can't be responsible for any terminal's response, but we are responsible to ensure that we're as compatible as possible and that we're passing whatever the compatible sequence is accurately. I feel like I'm preaching to the choir because I'm sure you know all this. 

> I'm also looking into optionally running combinations ...

For manual tests, I think it's a great idea to have some way of testing on a variety of operating systems and terminals. I don't really see that working in CI though, like you're saying. 

We could have some test code, that takes a file as a parameter, and in that file is a full path to the terminal(s) we want to test. Maybe even as an example or some type of test that doesn't run in CI but we can invoke manually. For these type of things, we probably need some manually invoked functionality that is different from CI where we just parse and check for the right escape codes.

fdncred replied:
> We could have some test code, that takes a file as a parameter, ...

In the shower today, I realized that this suggestion was crazy-talk.

> For manual tests, I think it's a great idea to have some way of testing on a variety of operating systems and terminals. 

I still think it would be good to have an example that we can run with `cargo run --example manual_test --features=whatev` that we can manually run at will in a variety of terminals to see if the title changes and whatnot.

> So it looks like CI is limited ...

If nothing else is possible for CI, then that's fine. The only reason I'm very interested in tests is because many people depend on this crate. I'd like to get to 48.1 or 49.0 sooner rather than later to avoid the breaking change in 48.

mhelsley replied:
@fdncred 
> > For manual tests, I think it's a great idea to have some way of testing on a variety of operating systems and terminals.
> 
> I still think it would be good to have an example that we can run with `cargo run --example manual_test --features=whatev` that we can manually run at will in a variety of terminals to see if the title changes and whatnot.

Yup, that's what the examples I created do except without the `--features=whatev` part:
```
cargo run --example hyperlink
cargo run --example title
...
```

Take a look at the last commit in the PR as it stands now. Just like the `cargo run --example 256_colors` CI step it adds the `OSC` examples to the CI runs and you can see the output in the most recent CI run:
```
Run cargo run --example hyperlink
   Compiling nu-ansi-term v0.48.0 (D:\a\nu-ansi-term\nu-ansi-term)
    Finished dev [unoptimized + debuginfo] target(s) in 0.26s
     Running `target\debug\examples\hyperlink.exe`

```

Notice how the text `Link to example.com` is missing. I think that just shows the CI output processing code doesn't know what to make of the hyperlink output but it does notice the Blue + underline styling.

> 
> > So it looks like CI is limited ...
> 
> If nothing else is possible for CI, then that's fine. The only reason I'm very interested in tests is because many people depend on this crate. I'd like to get to 48.1 or 49.0 sooner rather than later to avoid the breaking change in 48.

Makes sense. There are so many people using various seemingly-obscure open source projects that don't show up in various metrics that it's unsafe to assume the few people use it. Sometimes things being quiet just means folks are happy with it and that's a good "problem" to have.

It's probably more much more involved and speculative but perhaps there are some container CI + screenshot output workflows out there I could look into.

It's almost like we want a headless terminal (xtermjs? DomTerm?) somewhat like what phantomjs used to provide for web browser code.

I'll keep looking at those two -- running in container + screenshot or browser-based terminals but it would be really nice to get some of this merged soon. For instance, that Style test commit could be cherry-picked even without the rest (though I suppose it would show CI failing until the `OSC` fixes go in. Still, it better reflects the reality of the tree as it stands now..).

Alternately: Would  it make sense to put the `OSC` API behind a feature flag? I'm not usually a huge fan of APIs changing based on a feature flags because those can proliferate to the point of being their own special config / dll / registry hell. But it is fairly distinct and a secondary feature compared to the Styling and Coloring nu-ansi-term focuses on so I'm ambivalent on this idea.

fdncred replied:
> Yup, that's what the examples I created do ...

I'll try to look at them tonight sometime.

> It's almost like we want a headless terminal (xtermjs? DomTerm?) somewhat like what phantomjs used to provide for web browser code.

Have you looked at Reilly's [escape artist](https://github.com/rgwood/escape-artist)? You might something interesting in there,specifically around vte.

> Alternately: Would it make sense to put the OSC API behind a feature flag?

I'm not sure there's a lot of value there because there are so few commands that would be behind that feature. If we can do what you're doing here, without breaking everyone else, I think we'll be fine but I'm interested what others think (if they're watching this thread).

mhelsley replied:
OK, so I looked at xtermjs, escape artist, and libvte. The latter two would require taking screenshots and couldn't do a simple pass/fail in CI. Taking the screenshots is still not something I found a way to do nicely.

xtermjs has a headless mode and API. It can take input and use that to set its window and icon title fields, for example, which we could then check to see if they match the input we gave it.

However, xtermjs requires a slew of new code and new CI bits which will each need their own testing to become reliable. On the plus side though, whenever the CI stuff breaks fewer people will notice.

Then I found vtparse (https://github.com/haberman/vtparse). It's an incredibly tiny table-driven parser for virtual terminal input that uses Ruby to generate C code from their state machine description. It builds a library (.a) very easily and has only two function entry points -- initialization and consuming a byte of input. In turn it invokes a supplied callback indicating what's happening (e.g. OSC sequence start, OSC sequence end). It parses CSI and OSC sequences so it'd be useful for testing coloring, styling, and OSC sequences. So I've been working on a rust wrapper for it.

mhelsley commented:
@fdncred  To save you time on reviewing: Added another commit which introduces test code for detecting similar breaking changes in the future.

sholderbach commented on src/display.rs:
```diff
@@ -109,25 +134,44 @@ impl<'a, S: 'a + ToOwned + ?Sized> AnsiGenericString<'a, S>
+        &mut self.style
+    }
+
+    // Directly access the underlying string
+    pub fn as_str(&self) -> &S {
+        self.string.as_ref()
+    }
```
Can you make this a doc comment?

```suggestion
    /// Directly access the underlying string
    pub fn as_str(&self) -> &S {
        self.string.as_ref()
    }
```

(I know you just moved this up here, but worth the cleanup)

mhelsley replied:
@sholderbach 
> Can you make this a doc comment?

Done.

sholderbach replied:
Thx!

fdncred commented:
I'm trying to test the examples. Here's my results.

## CWD

![image](https://github.com/nushell/nu-ansi-term/assets/343840/a6c8d641-906b-4609-b20f-b86906088546)
And while it's waiting, I start a new terminal window, which I think is supposed to start in the `for_cwd` folder, but it doesn't. Maybe I'm misunderstanding what `cwd` is supposed to do.

## Hyperlink

Hyperlink doesn't look right either, but when I click on the hyperlink it goes to example.com
![image](https://github.com/nushell/nu-ansi-term/assets/343840/f5ab55aa-65d2-4061-a56c-882ceda6b6e1)

## icon

icon may just not work in Windows Terminal (not sure if it's supported) but it does change like this. the printout looks wrong as in the others.
![image](https://github.com/nushell/nu-ansi-term/assets/343840/2643d868-2977-4fcd-9199-2fe32da9bace)

![image](https://github.com/nushell/nu-ansi-term/assets/343840/fc7e94f2-c62d-4a50-8ff8-f6f51eac3470)

## title
seems to be working, except for what it's printing out
![image](https://github.com/nushell/nu-ansi-term/assets/343840/0fc5dd6e-5e6d-4593-bbfb-dfdafcd5e306)

![image](https://github.com/nushell/nu-ansi-term/assets/343840/7ff7d37a-d1df-4807-ad28-4bb8919fd365)

fdncred commented:
It's the `\x01` an `\x02` that are messing things up on Windows. I removed them and here are the results for hyperlink.
![image](https://github.com/nushell/nu-ansi-term/assets/343840/03c984f4-0712-4e15-9029-b6d7028a09a7)

I ran all the examples and the output on the screen looked right. I'm still not sure how to see what `cwd` does.

mhelsley commented:
> It's the `\x01` an `\x02` that are messing things up on Windows. I removed them and here are the results for hyperlink. ![image](https://user-images.githubusercontent.com/343840/246863453-03c984f4-0712-4e15-9029-b6d7028a09a7.png)

Oh, interesting. OK, I'll remove them from each one then.

Can you try putting a hyperlink in your command prompt? (e.g. setting one in PS1 in bash)
The reason I put those in is bash and readline use them to determine the invisible portions of the prompt so it knows how long the prompt is visually and can avoid clobbering the prompt and uses the full number of columns available when editting, choosing history, etc.

> I ran all the examples and the output on the screen looked right. I'm still not sure how to see what `cwd` does.

If you've got a Mac try running it in Terminal.app and, while it's waiting 5 seconds, try opening a new window. It should open the new terminal window in the ${PWD}/for_cwd/ directory rather than the ${PWD} directory you're running the example from (e.g. the top directory of the repo).

fdncred commented:
> Can you try putting a hyperlink in your command prompt? (e.g. setting one in PS1 in bash)

This would take some work. I'd have to compile nushell with this PR, write a nushell command to respond to your new hyperlink command, and call it from my script. Is there maybe an easier way to do this? Maybe I can just tweak the example and call it?

> If you've got a Mac try running it in Terminal.app and, while it's waiting 5 seconds, try opening a new window (it's been a while since I've used Terminal.app since I no longer have a Mac -- if there's a way to do it that's associated with that original window then that's the way you want to try). It should open the new terminal window in the ${PWD}/for_cwd/ directory rather than the ${PWD} directory you're running the example from (e.g. the top directory of the repo).

I tried iterm2, wezterm, and terminal.app and couldn't get any of them to respond properly to cwd

fdncred commented:
I put it in the right prompt and it looks like this on Windows 11 with Windows Terminal
![image](https://github.com/nushell/nu-ansi-term/assets/343840/f03e6fcc-6ed6-4e6e-b4c7-defc48e1035e)

Tabby Windows
![image](https://github.com/nushell/nu-ansi-term/assets/343840/21d04403-d5af-41ff-877f-33c3409ab2f1)

WezTerm Windows
![image](https://github.com/nushell/nu-ansi-term/assets/343840/d361f762-aaba-4a22-bd07-837190ee9454)

Alacritty Windows
![image](https://github.com/nushell/nu-ansi-term/assets/343840/21d40279-7328-4bcb-a105-ea905c4e88eb)

mhelsley commented:
> > Can you try putting a hyperlink in your command prompt? (e.g. setting one in PS1 in bash)
> 
> This would take some work. I'd have to compile nushell with this PR, write a nushell command to respond to your new hyperlink command, and call it from my script. Is there maybe an easier way to do this? Maybe I can just tweak the example and call it?

You could modify the hyperlink example to only output the hyperlink (no other content in the format string) and not sleep.

Then, at least in a new bash shell (e.g. run bash from within your current nushell) you could do:
```
unset PROMPT_COMMAND
PS1=$(./example/hyperlink)
```

If you want to manually write the hyperlinks without nu-ansi-term here's how to try two variations. For the variation with the \0x01..\x02 wrapper (which prevents command line editing/history navigation errors in bash at least):
```
unset PROMPT_COMMAND
PS1=$(echo -e '\\[\e]8;;https://example.com\e\\\\\]link example\\[\e]8;;\e\\\\\\] $ ')
```

Without the \0x01..\x02 wrapper (should create weird problems modifying command lines, navigating history with up-arrow, etc.):
```
unset PROMPT_COMMAND
PS1=$(echo -e '\e]8;;https://example.com\e\\\\link example\e]8;;\e\\\\ $ ')
```

and that will put a hyperlink in your prompt. I don't know how to make nushell invoke a program to get prompt content but that's essentially what you're looking for.

> > If you've got a Mac try running it in Terminal.app and, while it's waiting 5 seconds, try opening a new window (it's been a while since I've used Terminal.app since I no longer have a Mac -- if there's a way to do it that's associated with that original window then that's the way you want to try). It should open the new terminal window in the ${PWD}/for_cwd/ directory rather than the ${PWD} directory you're running the example from (e.g. the top directory of the repo).
> 
> I tried iterm2, wezterm, and terminal.app and couldn't get any of them to respond properly to cwd

Odd. Well, maybe I ought to remove `::cwd()` since we can't find a single terminal that seems to do anything with it.

mhelsley commented:
> I put it in the right prompt and it looks like this on Windows 11 with Windows Terminal ![image](https://user-images.githubusercontent.com/343840/246935105-f03e6fcc-6ed6-4e6e-b4c7-defc48e1035e.png)

Looks like this is the only one so far that doesn't like the `\x01 ... \x02` sequences.

mhelsley commented:
[vtparse](https://github.com/mhelsley/vtparse:master) the CI package I've been working on which could help test all kinds of Color, Style, and OSControl pieces.

fdncred commented:
I think we should remove the `\x01` and `\x02` as the default way to do this but document it somehow for people that need it. If people want to use is they can add the SOH/STX manually.

Another option is to add a parameter to the function or a separate function to add the SOH/STX bookends.

fdncred commented:
> testing

I love the examples and don't mind running them manually from time to time. i haven't tried running them in ci. however, i've not been able to make cwd work at all.

> SOH/STX

Ok, I'm good with another param or separate function. I'll let you decide which is best.

mhelsley commented:
Almost there. Pushed some changes so you can see, in a separate commit rather than with the fix, the "wrapping" code.

I do need to check the `ESC ST` sequence because I think there's something I'm not quite understanding there. If I don't output two `\\\\` when it's in the output of a PROMPT_COMMAND in bash then there are spurious `[` characters and command line editing is somehow worse.

I've also played with wrapping all of the SGR sequences. The code is much more complicated, though, and I think that would have to be a separate PR.

There's a TODO REVIEW in "Add optional OSC test combinatorial run" which I could use some help with; I don't know what the command line needs to be for Windows Terminal, WezTerm, etc. -- basically the non-Linux ones. I don't want to leave it as a TODO if I can avoid it.

fdncred commented:
> "Add optional OSC test combinatorial run"

I took the `#[ignore]` off the tests and tried to run them and they all passed. I'm not sure how since I don't have all those terminals on my mac. Maybe I'm just using the wrong command. I tried `cargo test osc` and `cargo test --all-features`.

The rest is looking good.

mhelsley commented:
## Running the (Ignored) Combinatorial Tests
> > "Add optional OSC test combinatorial run"
>
> I took the `#[ignore]` off the tests and tried to run them and they all passed. I'm not sure how since I don't have all those terminals on my mac. Maybe I'm just using the wrong command. 

Because it doesn't consider absence of the terminal to be a failure. That said, "success" doesn't seem like the right response either. I looked into how to report a "skip" at runtime and everything I've seen suggests there's no real way to do so *via test summary output* -- either it's statically ignored, filtered out by the command line, or it produces a pass/fail result.

So I modified the test to useful print output using `eprintln!()`. It will indicate whether the terminal is "Missing" or "Platform cannot run". However, unlike the default summary, to see the useful output you have to run the tests:

```
$ cargo test -- --ignored --show-output
...
$ cargo test --all-features -- --ignored --show-output
...
```
Of course, since it tries to run GUI terminals, that's not useful to put that in the CI runs. But if you've got Terminal.exe or Terminal.app it should open a new one for each OSC function. If it doesn't then the code for launching those commands isn't right and that's why it's still got the REVIEW comment there.

Note: For some reason cargo test is no longer letting me run only the `OSC` tests in `tests/osc.rs` -- no matter what I do it just ignores them. I've given up trying to discern why since the `--ignored` flag does effectively the same thing at the moment.

## Combinatorial Test Reorganization
I've reorganized the combinatorial test code. Now there will be one test ok/fail output per combination of `OSC` function to test and terminal to try. This more accurately reflects what's going on rather that hiding them by looping in `run_combos()`. It was useful because it could help me see that gnome-terminal wasn't being run even though I have it. So that's fixed too. The test functions will now show up in the summary as, for example, `hyperlink::XTerm()`.

## Dropping the Wrapping Stuff
Lastly, and most importantly, I figured out the issue I was seeing with needing to output two `\` for `ST`. In prompt strings it needs to be escaped *in addition* to the `0x01 ... 0x02` wrapping. Seeing as that adds a whole other wrinkle on top of the wrapping bits, I'd still need to modify Starship to do that, etc. I finally realized: the wrapping code is an added feature which probably still needs some work. As such, rather than rolling it into a fix it should be a separate PR -- possibly including some handy escaping bits -- rather than delay this. So I moved it to my personal `wrap-all` branch. I think this is the right move but if you disagree I can bring it back for this PR.

fdncred commented:
I was wondering why Terminal.app was failing on my Mac. I found out that it's because the path is wrong. I'm not sure if this has changed in MacOS or it's just a M1 and later change? This is the new path.
```
/System/Applications/Utilities/Terminal.app/Contents/MacOS/Terminal
```

mhelsley commented:
> I was wondering why Terminal.app was failing on my Mac. I found out that it's because the path is wrong. I'm not sure if this has changed in MacOS or it's just a M1 and later change? This is the new path.
> 
> ```
> /System/Applications/Utilities/Terminal.app/Contents/MacOS/Terminal
> ```

OK, I've replaced the path in my local tree. However in light of the rest of your reply I've kept digging.

> But it just hangs now and no terminal.app is shown. Not sure what's going on exactly.
> 
> Oh, this is interesting. I get this.
> 
> ```
> running 24 tests
> ...
> test cwd::TerminalApp has been running for over 60 seconds
> test hyperlink::TerminalApp has been running for over 60 seconds
> test icon::TerminalApp has been running for over 60 seconds
> test title::TerminalApp has been running for over 60 seconds
> ```

Interesting. I didn't see any command line arguments that could be supplied to potentially fix that. However I've found [another reference](https://ss64.com/osx/open.html) which may be potentially better or even fix your issue with the terminals not showing up. By any chance does:
```
open -n -F -W -a "/Applications/Utilities/Terminal.app" --args /usr/bin/sleep 10
```

open a terminal? The `-W` flag says to wait for the command to complete and, since it should be running `sleep 10` it should finish waiting in 10 seconds so hopefully you won't see the same problem with lingering processes. The `-a` flag looks like it's supposed to supply a path to a `.app` in the `/Applications` directory however you could also try it with `/System/` prepended if that fails.

How about:
```
open -n -F -W -b "com.apple.terminal" --args /usr/bin/sleep 10
```

This one replaces a path to the application with what I believe is a "bundle" reference of some kind.

> 
> But if i do a `ps` in another window, I see this
> 
> ```
> 46791 ttys000    0:00.08 /Users/fdncred/.rustup/toolchains/stable-aarch64-apple-darwin/bin/cargo test -- --ignored --show-output
> 46835 ttys000    0:00.01 /Users/fdncred/src/nu-ansi-term/target/debug/deps/osc-98ace87dd48bab95 --ignored --show-output
> 46836 ttys000    0:00.08 /System/Applications/Utilities/Terminal.app/Contents/MacOS/Terminal  ^[]7;./for_cwd/^[\
> 46837 ttys000    0:00.08 /System/Applications/Utilities/Terminal.app/Contents/MacOS/Terminal  ^[[4;34m^[]8;;https://example.com^[\Link to example.com^[]8;
> 46838 ttys000    0:00.09 /System/Applications/Utilities/Terminal.app/Contents/MacOS/Terminal  ^[]I;./foo.icn^[\
> 46839 ttys000    0:00.08 /System/Applications/Utilities/Terminal.app/Contents/MacOS/Terminal  ^[]2;My Title^[\
> ```
> 
> I just never see the window. I have to `killall Terminal` to get the test to return at all.

I'm guessing the Terminal.app is open in the foreground when you run the tests, but if not then perhaps bringing it to the foreground or looking at its window list (in the menu bar if I recall) will show them.

> > cargo test osc
> 
> it runs fine here
> 
> ```
> cargo test osc                                                ✘ 101   3m 36s   1.70.0  08:12:01 AM
>     Finished test [unoptimized + debuginfo] target(s) in 0.01s
>      Running unittests src/lib.rs (target/debug/deps/nu_ansi_term-7bfff3812bd5bd6a)
> 
> running 0 tests
> 
> test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 70 filtered out; finished in 0.00s

While I expected to see `70 filtered out` here because these are the non-osc tests...

> 
>      Running tests/osc.rs (target/debug/deps/osc-98ace87dd48bab95)
> 
> running 0 tests
> 
> test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 24 filtered out; finished in 0.00s

It *also* `filtered out` the 24 `osc` tests here. That's the source of my confusion. Perhaps every `#[test]`-attributed function needs to be prefixed by the string `osc` for it to recognize them? Or maybe they have to be directly under the `osc` crate/module -- they can't be in their own mod?

mhelsley commented:
> They were 4 terminal.apps open in my dock (which was hidden). Good guess. If i go to them and exit and quit the app, the tests will proceed.

Yeah, that seems to be normal. As best I can tell `cargo` runs a limited number of test cases in parallel (seems to be about 4-6 -- perhaps some proportion of the number of CPUs). The test isn't supposed to finish until all the windows are closed because that ensures you get a chance to see the displayed terminal. So it makes sense that it wouldn't finish until all the new windows are closed.

> However I can't tell that it's doing anything. i.e. the title isn't changed, there's no hyperlink, cwd doesn't seem to work, don't notice any icon. I'm know cwd works in terminal.app because I implemented it in nushell. So does the title. Not sure about icon. I don't think hyperlink works at all. We have a command in nushell called `ansi link` and I can't activate or see any link in terminal.app.

OK, so I managed to spend some time on a call with a friend to try variations on `open` and he said it doesn't actually execute anything inside the new terminal even if you supply a command to run after `--args`. That would explain why nothing works in `Terminal.app`.

I think that means, for now, it's best to omit MacOS and Terminal.app from the `tests/osc.rs` code. Since users can manually run the examples from within an already running Terminal via `cargo run --example foo` and Terminal is the only one the `osc.rs` code tries on Mac OS it's kind of pointless to stress about getting it running on Mac OS.

I considered running it via Applescript but my friend had to end our call. So I can only speculate on what the script might look like based on [this StackExchange](https://apple.stackexchange.com/a/170604). The script file (e.g. `myscript`) would have something like:
```
tell application "Terminal"
    do script "/bin/zsh -c cargo run --example hyperlink"
end tell
```
Which could be run (from a command line or via the `osc.rs` code) with something like:
```
osascript -e myscript
```

In the next few days I'll polish the PR up some more and push hopefully the last changes before it's ready for you to run the tests again and, if that passes your standards, a decision on merging.

Thanks for your time!

fdncred commented:
Testing on Windows this morning, trying to get things to work. Here's my current changes that aren't working exactly.

```rust
    term!(
        TerminalExe,
        target_family = "windows",
        "wt.exe"
    );
```
and
```rust
            #[cfg($plat)]
            fn exists() -> Option<bool> {
                //Some(is_executable($cmd))
                Some(true)
            }
```
On Windows, for `wt.exe`, we really just need a check to see if it's in the path versus at a particular location. I believe it changes location based on how it's installed and what version is installed. For example, this is the current location of mine `C:\Program Files\WindowsApps\Microsoft.WindowsTerminal_1.17.11461.0_x64__8wekyb3d8bbwe\wt.exe`.

Also, I haven't figured out how to pass commands to it programmatically. I get windows like this with test command

![image](https://github.com/nushell/nu-ansi-term/assets/343840/51f53258-d8ae-404c-afdb-560f30045c9a)

This is the help for wt.exe
```
---------------------------
Help
---------------------------
wt - the Windows Terminal
Usage: [OPTIONS] [SUBCOMMAND]

Options:
  -h,--help                   Print this help message and exit
  -v,--version                Display the application version
  -M,--maximized Excludes: --fullscreen
                              Launch the window maximized
  -F,--fullscreen Excludes: --focus --maximized
                              Launch the window in fullscreen mode
  -f,--focus Excludes: --fullscreen
                              Launch the window in focus mode
  --pos TEXT                  Specify the position for the terminal, in "x,y" format.
  --size TEXT                 Specify the number of columns and rows for the terminal, in "c,r" format.
  -w,--window TEXT            Specify a terminal window to run the given commandline in. "0" always refers to the current window. 
  -s,--saved INT              This parameter is an internal implementation detail and should not be used.

Subcommands:
  new-tab                     Create a new tab
  nt                          An alias for the "new-tab" subcommand.
  split-pane                  Create a new split pane
  sp                          An alias for the "split-pane" subcommand.
  focus-tab                   Move focus to another tab
  ft                          An alias for the "focus-tab" subcommand.
  move-focus                  Move focus to the adjacent pane in the specified direction
  mf                          An alias for the "move-focus" subcommand.
  move-pane                   Move focused pane to another tab
  mp                          An alias for the "move-pane" subcommand.
  swap-pane                   Swap the focused pane with the adjacent pane in the specified direction
  focus-pane                  Move focus to another pane
  fp                          An alias for the "focus-pane" subcommand.


---------------------------
OK   
---------------------------
```

If I run it manually like this
![image](https://github.com/nushell/nu-ansi-term/assets/343840/0d419629-2c78-4381-a96a-537fe6942688)
I get these windows and they disappear after 10 seconds
![image](https://github.com/nushell/nu-ansi-term/assets/343840/3c272330-8748-4ca5-9ecb-654d94fcd73f)
![image](https://github.com/nushell/nu-ansi-term/assets/343840/2e2a0384-5f11-4b67-abd7-b3b78e149d61)
![image](https://github.com/nushell/nu-ansi-term/assets/343840/e123416d-5c1f-490b-8317-909a50062359)
![image](https://github.com/nushell/nu-ansi-term/assets/343840/649a7a12-2903-4ea5-b1e8-d438f02c096c)

fdncred commented:
I'm fine with running examples manually, but we need to be able to see them work. I can't see `cwd` or `icon` doing anything. If we can't get them to produce some result, I think we should probably remove them.

Running on WSL2 I have
gnome-terminal - title and hyperlink work, can't tell on cwd and icon
rxvt - same
eterm - couldn't get to work

i haven't tried changing the code for these other terminal emulators but it's some i've heard of
st
foot
kitty
wezterm
alacritty
tabby

mhelsley commented:
> I'm fine with running examples manually, but we need to be able to see them work. I can't see `cwd` or `icon` doing anything. If we can't get them to produce some result, I think we should probably remove them.

That makes sense.

## TL;DR / Summary:
* Remove `::cwd()`  (I'll make this it's own commit)
* Rename `::icon()` to something more accurate (e.g. `::icon_caption()` or `::icon_title()`) or just Remove it too?

I'm a little hesitant to remove `::icon()` because I could see it being useful *somewhere*. However, if you say to remove it I'll also happily do that since it means less to fix/maintain. If you want to know more detail why I hesitate there's an optional boring bit at the bottom of this comment.

> Running on WSL2 I have gnome-terminal - title and hyperlink work, can't tell on cwd and icon rxvt - same eterm - couldn't get to work

Yeah, I'm having trouble with eterm as well. If I can't get it working locally and soon I'll just remove it. I think we've got enough confidence from the existing coverage and adding more terminals (on Linux) begins to have diminishing returns.

> i haven't tried changing the code for these other terminal emulators but it's some i've heard of st foot kitty wezterm alacritty tabby

I've tried `alacritty` in the past. The issue with adding `alacritty` and most of the less common terminal emulators to the testing is how reliably Rust's `std::process::Command` would find/detect them on all the platforms. That said, I will modify the `OSC` test code to make it less brittle in that respect and other terminals can be added (or removed) more easily in future PRs.

----
####  Possible Origins of `::icon()` Confusion (skip unless supremely bored)
My guess is tech cultural terseness and history resulted in my confusion at least. I now think the "icon" feature has *very little to do with icons*. I seem to recall in old school X Windows,when you "minimized" a window, it became an "icon" on your "root window" (modern: "the desktop"). It helped users choose among windows represented by the same *icons* they wanted to unminimize. But for its origins with X Windows and xterm, it might better have been described as a "brief caption" or "name" for whatever is currently running in that window rather than having anything to do with an "icon".

In the intervening years, since that UX has largely disappeared, it's become synonymous in many terminal emulators with setting the window title. It's hopelessly confused and now `OSC` 0, I, and 2 all seem to do exactly the same thing *at least in practice*. Chances are only the oldest terminal emulators (perhaps xterm) make any subtle distinction between the two.

fdncred commented:
> when you "minimized" a window, it became an "icon" on your "root window" (modern: "the desktop")

ah, yes. I remember this. I run SunOS 4.1.3/4 for many years and remember this.

> I'll also happily do that since ...

ya, let's remove cwd and icon please.

Let me know when you're ready for more testing.

mhelsley commented:
> ```rust
>     term!(
> ...
>         "wt.exe"
>     );
> ...
>             #[cfg($plat)]
>             fn exists() -> Option<bool> {
> ```

OK, I've merged those changes in and simplified things a lot. I renamed the function `::exists() -> Option<bool>` to `::try() -> bool` and it only passes along the platform `cfg`. Then `std::process::Command` does the searching and checking to see if the terminal can be executed. Only if it finds and executes the terminal can the test case possibly fail. I think that will deal with the problem you described:

> On Windows, for `wt.exe`, we really just need a check to see if it's in the path versus at a particular location. I believe it changes location based on how it's installed and what version is installed. For example, this is the current location of mine `C:\Program Files\WindowsApps\Microsoft.WindowsTerminal_1.17.11461.0_x64__8wekyb3d8bbwe\wt.exe`.
 
As for:
> Also, I haven't figured out how to pass commands to it programmatically. I get windows like this with `cargo test --all-features -- --ignored --show-output`.
> 
> ![image](https://user-images.githubusercontent.com/343840/249156183-51f53258-d8ae-404c-afdb-560f30045c9a.png)

The inner double-quotes showing up inside the backticks suggests the double-quotes might be interfering with parsing the command line properly -- I'm guessing that `wt.exe` is adding the outer backticks around the input it received. Is there some way to submit the cargo command portion using a different combination of or perhaps even no quoting?

Since I lack access to Windows, by any chance could I submit this PR and you could fix up the Windows bits? I'm happy to respond with guesses and possible solutions for problems you encounter. I just think we've got these longer cycle times and it might become error-prone to have me modifying the PR based on your descriptions each time. That might work better unless I'm mistaken and you're unable to modify the PR too.

> This is the help for wt.exe
> ```
> wt - the Windows Terminal
> Usage: [OPTIONS] [SUBCOMMAND]
> 
> Options:
... (I didn't see any additional promising options to use)
> 
> Subcommands:
>   new-tab                     Create a new tab
>   nt                          An alias for the "new-tab" subcommand.

The `nt` subcommand looks like a potential way to isolate the test from potentially messing with whatever the user is already doing in another terminal window / tab.

> If I run it manually like this ... I get these windows and they disappear after 10 seconds !

OK, that looks good and I think that's sort of the ideal result (though maybe 10 seconds is too long / too short for some folks -- you folks can modify the timeout as you see fit before merging).

mhelsley commented:
> > when you "minimized" a window, it became an "icon" on your "root window" (modern: "the desktop")
> 
> ah, yes. I remember this. I run SunOS 4.1.3/4 for many years and remember this.
> 
> > I'll also happily do that since ...
> 
> ya, let's remove cwd and icon please.
> 
> Let me know when you're ready for more testing.

OK, I removed `::icon()` and `::cwd()`. I think it's ready for testing. I've left `TODO` notes in `tests/osc.rs` around the Windows and Mac testing portions.

I believe I've done everything I planned on doing for this PR. Let me know if you see more that needs to be fixed or I need to do something for the Mac and Windows paths in `tests/osc.rs`.

fdncred commented on src/display.rs:
```diff
@@ -310,13 +300,38 @@ where
+                w.write_str(self.string.as_ref())?;
+                write!(w, "\x1B\x5C")
+            }
+            Some(OSControl::Icon) => {
```
I think we should remove the Icon and Cwd parts.

mhelsley replied:
Argh, yeah I missed that.

mhelsley replied:
OK, pushed again. Unlike last time I grep'd for `icon` and `cwd` so that should do it.

fdncred commented:
@mhelsley I'm reconsidering the value of the osc.rs test. I mean, it's cool but if we can't get it to work cross platform, i'm not sure we should leave it in. I think I'd probably land this now if you can remove it. What are your thoughts on it?

mhelsley commented:
> @mhelsley I'm reconsidering the value of the osc.rs test. I mean, it's cool but if we can't get it to work cross platform, i'm not sure we should leave it in. I think I'd probably land this now if you can remove it. What are your thoughts on it?

It makes sense that the testing should be cross platform. I've moved the code into it's own  local branch for now. I also merged your "remove comments" change into the commit with the examples.

fdncred commented:
Thanks. Let's move forward with this!
